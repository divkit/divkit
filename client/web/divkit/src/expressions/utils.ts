import type { EvalContext, EvalValue, EvalValueBase } from './eval';
import type { Node, Variable } from './ast';
import type { EvalTypes } from './eval';
import type { VariablesMap } from './eval';
import { walk } from './walk';
import { parseColor, ParsedColor } from '../utils/correctColor';
import { padLeft } from '../utils/padLeft';
import { hasBigInt, MAX_INT, MIN_INT, toBigInt } from './bigint';

export function valToInternal(val: EvalValue): EvalValue {
    if (val.type === 'url' || val.type === 'color') {
        return {
            type: 'string',
            value: val.value
        };
    }

    return val;
}

export function dateToString(date: Date): string {
    return date
        .toISOString()
        .replace('T', ' ')
        .replace(/\.\d{3}Z$/, '');
}

export function valToString(val: EvalValue): string {
    if (val.type === 'string') {
        return val.value;
    } else if (val.type === 'integer') {
        return String(val.value);
    } else if (val.type === 'number') {
        let res = String(val.value);

        if (!res.includes('.')) {
            if (res.includes('e')) {
                res = res.replace('e', '.0e');
            } else {
                res += '.0';
            }
        }

        res = res.replace(/e\+?/i, 'E');

        return res;
    } else if (val.type === 'boolean') {
        return val.value ? 'true' : 'false';
    } else if (val.type === 'datetime') {
        return dateToString(val.value);
    } else if (val.type === 'color') {
        return stringifyColor(safeConvertColor(val.value));
    } else if (val.type === 'url') {
        return val.value;
    } else if (val.type === 'dict') {
        return '<dict>';
    }

    // For purpose when new eval value types will be added
    throw new Error(`Unexpected type ${(val as EvalValueBase).type}`);
}

export function valToPreview(val: EvalValue): string {
    let res = valToString(val);

    if (val.type === 'string') {
        res = "'" +
            res
                .replace(/\\/g, '\\\\')
                .replace(/'/g, '\\\'') +
            "'";
    }

    return res;
}

export function typeToString(type: EvalTypes): string {
    if (type === 'datetime') {
        return 'DateTime';
    }
    return type.charAt(0).toUpperCase() + type.substring(1);
}

export function roundInteger(ctx: EvalContext, val: number | bigint): number | bigint {
    if (hasBigInt) {
        return toBigInt(val);
    }

    if (val < 0) {
        const res = -Math.floor(-val as number);
        if (res === 0) {
            // -0 => 0
            return 0;
        }
        return res;
    }

    return Math.floor(val as number);
}

export function checkIntegerOverflow(ctx: EvalContext, val: number | bigint): void {
    if (val < MIN_INT || val > MAX_INT) {
        throw new Error('Integer overflow.');
    }

    if (!hasBigInt && (val > Number.MAX_SAFE_INTEGER || val < Number.MIN_SAFE_INTEGER)) {
        ctx.safeIntegerOverflow = true;
    }
}

export function gatherVarsFromAst(ast: Node): string[] {
    const res: string[] = [];

    walk(ast, {
        Variable(node: Variable): void {
            res.push(node.id.name);
        }
    });

    return res;
}

export function evalError(msg: string, details: string): never {
    throw new Error(`Failed to evaluate [${msg}]. ${details}`);
}

export function containsUnsetVariables(ast: Node, variables: VariablesMap): boolean {
    let result = false;

    walk(ast, {
        Variable(node: Variable): void {
            if (!variables.has(node.id.name)) {
                result = true;
            }
        }
    });

    return result;
}

export function safeConvertColor(color: string): ParsedColor {
    const res = parseColor(color);

    if (res) {
        return res;
    }

    throw new Error('Unable to convert value to Color, expected format #AARRGGBB.');
}

export function stringifyColor(color: ParsedColor): string {
    return `#${[color.a, color.r, color.g, color.b].map(it => {
        if (it < 0 || it > 255) {
            throw new Error('Value out of range 0..1.');
        }

        return padLeft(Math.round(it).toString(16), 2);
    }).join('').toUpperCase()}`;
}

export function transformColorValue(color: string): string {
    return stringifyColor(safeConvertColor(color));
}
