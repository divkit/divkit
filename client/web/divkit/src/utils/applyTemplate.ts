import { wrapError, type WrappedError } from './wrapError';
import type { DivBase, TemplateContext } from '../../typings/common';
import type { DivBaseData } from '../types/base';

export function applyTemplate<T extends DivBase>(
    json: T,
    templateContext: TemplateContext,
    templates: Record<string, unknown>,
    logError: (error: WrappedError) => void
): {
    json: T;
    templateContext: TemplateContext;
};
export function applyTemplate(
    json: DivBaseData,
    templateContext: TemplateContext,
    templates: Record<string, unknown>,
    logError: (error: WrappedError) => void
): {
    json: DivBaseData;
    templateContext: TemplateContext;
} {
    const template = templates[json.type];

    if (!template) {
        logError(wrapError(new Error('No such template'), {
            additional: {
                template: json.type
            }
        }));

        return {
            json,
            templateContext
        };
    }

    let i;
    const newContext: TemplateContext = {};

    for (i in templateContext) {
        if (templateContext.hasOwnProperty(i)) {
            newContext[i] = templateContext[i];
        }
    }

    for (i in json) {
        if (i === 'type' || i === '__proto__') {
            continue;
        }

        if (json.hasOwnProperty(i)) {
            newContext[i] = json[i as keyof typeof json];
        }
    }

    function copyTemplated(base: any, extender: any) {
        const keys = Object.keys(extender).filter(key => key !== '__proto__');
        const simpleKeys = keys.filter(key => key.charAt(0) !== '$');
        const templateKeys = keys.filter(key => key.charAt(0) === '$');

        simpleKeys.forEach(key => {
            const item = extender[key];

            if (typeof item === 'object' && item !== null) {
                base[key] = Array.isArray(item) ? [] : {};
                copyTemplated(base[key], item);
            } else {
                base[key] = item;
            }
        });

        templateKeys.forEach(key => {
            const item = extender[key];

            const val = newContext[item];

            if (val !== undefined) {
                const prop = key.substring(1);
                base[prop] = val;
            }
        });

        return base;
    }

    const newJson = copyTemplated({}, template);

    for (i in json) {
        if (i === 'type' || i === '__proto__') {
            continue;
        }

        if (json.hasOwnProperty(i)) {
            newJson[i] = json[i as keyof typeof json];
        }
    }

    return {
        json: newJson,
        templateContext: newContext
    };
}
