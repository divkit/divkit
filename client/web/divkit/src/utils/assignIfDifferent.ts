import { isDeepEqual } from './isDeepEqual';

export function assignIfDifferent<A, B>(newVal: A, defaultVal: B): A | B {
    if (isDeepEqual(newVal, defaultVal)) {
        return defaultVal;
    }
    return newVal;
}
