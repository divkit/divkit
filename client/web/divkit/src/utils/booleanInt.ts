import type { BooleanInt } from '../../typings/common';

export function booleanInt(val: BooleanInt): boolean {
    return val === true || val === 1;
}
