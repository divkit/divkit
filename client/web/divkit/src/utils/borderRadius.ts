import type { CornersRadius } from '../types/border';
import { pxToEm } from './pxToEm';

export function borderRadius(cornersRadius: CornersRadius, defaultRadius = 0, fontSize = 10): string {
    return [
        cornersRadius['top-left'],
        cornersRadius['top-right'],
        cornersRadius['bottom-right'],
        cornersRadius['bottom-left']
    ]
        .map(it => pxToEm((it || defaultRadius) / fontSize * 10))
        .join(' ');
}
