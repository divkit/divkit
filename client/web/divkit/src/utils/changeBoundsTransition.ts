import { cubicInOut } from 'svelte/easing';
import type { ChangeBoundsTransition } from '../types/base';
import { lerp } from './lerp';
import { isPrefersReducedMotion } from './isPrefersReducedMotion';
import { EASING } from './easing';

const DEFAULT_DURATION = 200;
const DEFAULT_DELAY = 0;

export interface TransitionProps {
    rootBbox: DOMRect;
    beforeBbox: DOMRect;
    afterBbox: DOMRect;
    transition: ChangeBoundsTransition;
}

export function changeBoundsTransition(node: HTMLElement, {
    rootBbox,
    beforeBbox,
    afterBbox,
    transition
}: TransitionProps) {
    return {
        delay: transition.start_delay ?? DEFAULT_DELAY,
        duration: isPrefersReducedMotion() ? 0 : (transition.duration ?? DEFAULT_DURATION),
        easing: (transition.interpolator && transition.interpolator in EASING) ?
            EASING[transition.interpolator] :
            cubicInOut,
        css: (t: number) => {
            return [
                `top:${lerp(beforeBbox.top, afterBbox.top, t) - rootBbox.top}px`,
                `left:${lerp(beforeBbox.left, afterBbox.left, t) - rootBbox.left}px`,
                `width:${lerp(beforeBbox.width, afterBbox.width, t)}px`,
                `height:${lerp(beforeBbox.height, afterBbox.height, t)}px`
            ].join(';');
        }
    };
}
