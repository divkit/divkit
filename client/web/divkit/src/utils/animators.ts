import type { AnimatorDirection, TypedValue } from '../../typings/common';
import type { MaybeMissing } from '../expressions/json';
import type { Animator } from '../types/base';
import type { Variable, VariableType } from '../../typings/variables';
import type { ExecAnyActionsFunc } from '../context/root';
import { stringifyColor } from '../expressions/utils';
import { clamp } from './clamp';
import { parseColor } from './correctColor';
import { correctNonNegativeNumber } from './correctNonNegativeNumber';
import { correctPositiveNumber } from './correctPositiveNumber';
import { getEasing } from './easing';
import { lerp } from './lerp';

function correctDirection(direction: string | undefined): AnimatorDirection | undefined {
    if (direction === 'normal' || direction === 'reverse' || direction === 'alternate' || direction === 'alternate_reverse') {
        return direction;
    }
}

export type CalcedAnimator = MaybeMissing<Animator & {
    start_value_typed: TypedValue;
    end_value_typed: TypedValue;
}>;

export interface AnimatorInstance {
    stop(): void;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createAnimator(
    animator: CalcedAnimator,
    variableInstance: Variable<any, VariableType>,
    onEnd: () => void,
    execAnyActions: ExecAnyActionsFunc
): AnimatorInstance | undefined {
    const duration = correctPositiveNumber(animator.duration, 0);

    if (!duration || animator.type !== 'color_animator' && animator.type !== 'number_animator') {
        return;
    }

    const startValue = (animator.start_value_typed ? animator.start_value_typed.value : animator.start_value) ??
        variableInstance.getValue();
    const endValue = animator.end_value_typed ? animator.end_value_typed.value : animator.end_value;

    if (startValue === undefined || endValue === undefined) {
        return;
    }

    if (
        animator.type === 'color_animator' && (typeof startValue !== 'string' && startValue !== undefined || typeof endValue !== 'string') ||
        animator.type === 'number_animator' && (typeof startValue !== 'number' && startValue !== undefined || typeof endValue !== 'number')
    ) {
        return;
    }

    const parsedStart = animator.type === 'color_animator' && parseColor(startValue as string);
    const parsedEnd = animator.type === 'color_animator' && parseColor(endValue as string);
    if (animator.type === 'color_animator' && (!parsedStart || !parsedEnd)) {
        return;
    }

    const delay = correctNonNegativeNumber(animator.start_delay, 0);
    const easing = getEasing(animator.interpolator || 'linear');
    const direction = correctDirection(animator.direction) || 'normal';
    // eslint-disable-next-line no-nested-ternary
    const repeatCount = animator.repeat_count?.type === 'infinity' ?
        Infinity :
        (animator.repeat_count?.type === 'fixed' ? correctNonNegativeNumber(animator.repeat_count?.value, 1) : 1);

    let animationTime = 0;
    let localTime = performance.now();
    const maxTime = repeatCount === Infinity ? Infinity : repeatCount * duration + delay;

    function mix(t: number): string | number {
        if (animator.type === 'color_animator') {
            if (!parsedStart || !parsedEnd) {
                throw new Error('Missing start/end value');
            }
            return stringifyColor({
                a: clamp(lerp(parsedStart.a, parsedEnd.a, t), 0, 255),
                r: clamp(lerp(parsedStart.r, parsedEnd.r, t), 0, 255),
                g: clamp(lerp(parsedStart.g, parsedEnd.g, t), 0, 255),
                b: clamp(lerp(parsedStart.b, parsedEnd.b, t), 0, 255)
            });
        }
        return lerp(startValue as number, endValue as number, t);
    }

    function tick(now: number) {
        const diff = now - localTime;
        localTime = now;

        animationTime += diff;

        if (animationTime >= delay) {
            let iterationNumber = Math.floor((animationTime - delay) / duration);

            let t = (animationTime - delay - iterationNumber * duration) / duration;

            if (iterationNumber >= repeatCount) {
                iterationNumber = repeatCount - 1;
                t = 1;
            }

            let dir: 'normal' | 'reverse';
            if (
                direction === 'normal' ||
                (direction === 'alternate' && iterationNumber % 2 === 0) ||
                direction === 'alternate_reverse' && iterationNumber % 2 === 1
            ) {
                dir = 'normal';
            } else {
                dir = 'reverse';
            }

            if (dir === 'reverse') {
                t = 1 - t;
            }

            const value = mix(easing(t));
            variableInstance.setValue(value);
        }

        if (animationTime < maxTime) {
            timer = requestAnimationFrame(tick);
        } else {
            onEnd();
            execAnyActions(animator.end_actions);
        }
    }

    let timer = requestAnimationFrame(tick);

    return {
        stop() {
            cancelAnimationFrame(timer);
            execAnyActions(animator.cancel_actions);
            execAnyActions(animator.end_actions);
        }
    };
}
