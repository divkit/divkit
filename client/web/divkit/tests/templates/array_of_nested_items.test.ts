import {
    describe,
    expect,
    test,
    vi
} from 'vitest';

import { applyTemplate } from '../../src/utils/applyTemplate';

describe('array_of_nested_items', () => {
    test('test_array_of_nested_items_with_item_template', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_with_item_template.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_of_nested_items_with_item_template_without_data', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_with_item_template_without_data.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_of_nested_items_with_item_template_without_item_required_property', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_with_item_template_without_item_required_property.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_of_nested_items_with_link_for_items', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_with_link_for_items.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_of_nested_items_with_link_for_items_with_item_template', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_with_link_for_items_with_item_template.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_of_nested_items_with_link_for_items_without_item_required_property', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_with_link_for_items_without_item_required_property.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_array_of_nested_items_without_item_required_property', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_without_item_required_property.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_array_of_nested_items_without_templates', () => {
        const json = require('../../../../../test_data/template_test_data/array_of_nested_items/test_array_of_nested_items_without_templates.json');
        const logError = vi.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });
});
