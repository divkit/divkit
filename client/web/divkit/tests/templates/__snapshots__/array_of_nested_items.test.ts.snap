// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array_of_nested_items > test_array_of_nested_items_with_item_template 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_item_template 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_item_template_without_data 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_item_template_without_data 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_item_template_without_item_required_property 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_item_template_without_item_required_property 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_link_for_items 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
  },
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_link_for_items 2`] = `[]`;

exports[`array_of_nested_items > test_array_of_nested_items_with_link_for_items_with_item_template 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "items_link": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {
    "items_link": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
  },
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_with_link_for_items_without_item_required_property 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
  },
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_without_item_required_property 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_without_item_required_property 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items > test_array_of_nested_items_without_templates 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items > test_array_of_nested_items_without_templates 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items test_array_of_nested_items_with_item_template 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items test_array_of_nested_items_with_item_template 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items test_array_of_nested_items_with_item_template_without_data 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items test_array_of_nested_items_with_item_template_without_data 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items test_array_of_nested_items_with_item_template_without_item_required_property 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items test_array_of_nested_items_with_item_template_without_item_required_property 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items test_array_of_nested_items_with_link_for_items 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
  },
}
`;

exports[`array_of_nested_items test_array_of_nested_items_with_link_for_items 2`] = `[]`;

exports[`array_of_nested_items test_array_of_nested_items_with_link_for_items_with_item_template 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "items_link": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {
    "items_link": [
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "link": "https://ya.ru",
          "type": "item_entity_template",
        },
        "property": "Some text",
      },
    ],
  },
}
`;

exports[`array_of_nested_items test_array_of_nested_items_with_link_for_items_without_item_required_property 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {
    "items_link": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
  },
}
`;

exports[`array_of_nested_items test_array_of_nested_items_without_item_required_property 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items test_array_of_nested_items_without_item_required_property 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;

exports[`array_of_nested_items test_array_of_nested_items_without_templates 1`] = `
{
  "json": {
    "items": [
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
      {
        "entity": {
          "property": {
            "value": "https://ya.ru",
          },
          "type": "entity_with_complex_property",
        },
        "property": "Some text",
      },
    ],
    "type": "entity_with_array_of_nested_items",
  },
  "templateContext": {},
}
`;

exports[`array_of_nested_items test_array_of_nested_items_without_templates 2`] = `
[
  [
    [Error: No such template],
  ],
]
`;
